### PR TITLE
feat: GlobalAuctionRebalanceExtension contract, tests & utils

### PR DESCRIPTION
**Pull request by snake-poison**
Adds a GlobalAuctionRebalanceExtension that is compatible with the DelegatedManager.

tests are at 100% and it should be ready to serve as the base for a GlobalOptimisticAuctionExtension.

### CODE REVIEW

File: contracts/global-extensions/GlobalAuctionRebalanceExtension.sol [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+
+    /* ============ Structs ============ */
+
+    struct AuctionExecutionParams {
+        uint256 targetUnit;                      // Target quantity of the component in Set, in precise units (10 ** 18).
+        string priceAdapterName;                 // Identifier for the price adapter to be used.
+        bytes priceAdapterConfigData;            // Encoded data for configuring the chosen price adapter.
+    }
+
```
Comment (pblivin0x) [inline contracts/global-extensions/GlobalAuctionRebalanceExtension.sol:59]:
```suggestion

    /* ============ State Variables ============ */

```

File: contracts/global-extensions/GlobalAuctionRebalanceExtension.sol [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+ * @author Index Coop
+ *
+ * @dev Extension contract for interacting with the AuctionRebalanceModuleV1. This contract acts as a pass-through and functions 
+ * are only callable by operator. 
+ */
+contract GlobalAuctionRebalanceExtension is BaseGlobalExtension {
+    using AddressArrayUtils for address[];
+    using SafeMath for uint256;
+
+        /* ============ Events ============ */
```
Comment (pblivin0x) [inline contracts/global-extensions/GlobalAuctionRebalanceExtension.sol:44]:
```suggestion
    /* ============ Events ============ */
```

File: contracts/global-extensions/GlobalAuctionRebalanceExtension.sol [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        _initializeModule(_delegatedManager.setToken(), _delegatedManager);
+    }
+
+    /**
+     * @dev ONLY OWNER: Initializes AuctionRebalanceExtension to the DelegatedManager.
+     *
+     * @param _delegatedManager     Instance of the DelegatedManager to initialize
+     */
+    function initializeExtension(IDelegatedManager _delegatedManager) external onlyOwnerAndValidManager(_delegatedManager) {
+        ISetToken setToken = _delegatedManager.setToken();
```
Comment (pblivin0x) [inline contracts/global-extensions/GlobalAuctionRebalanceExtension.sol:87]:
```suggestion
        require(_delegatedManager.isPendingExtension(address(this)), "Extension must be pending");

        ISetToken setToken = _delegatedManager.setToken();
```

File: contracts/global-extensions/GlobalAuctionRebalanceExtension.sol [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    /* ============ External Functions ============ */
+
+    
+    /**
+     * @dev ONLY OWNER: Initializes AuctionRebalanceModuleV1 on the SetToken associated with the DelegatedManager.
+     *
+     * @param _delegatedManager     Instance of the DelegatedManager to initialize the AuctionRebalanceModuleV1 for
+     */
+    function initializeModule(IDelegatedManager _delegatedManager) external onlyOwnerAndValidManager(_delegatedManager) {
+        _initializeModule(_delegatedManager.setToken(), _delegatedManager);
```
Comment (pblivin0x) [inline contracts/global-extensions/GlobalAuctionRebalanceExtension.sol:78]:
```suggestion
        require(_delegatedManager.isInitializedExtension(address(this)), "Extension must be initialized");

        _initializeModule(_delegatedManager.setToken(), _delegatedManager);
```

File: contracts/global-extensions/GlobalAuctionRebalanceExtension.sol [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        emit AuctionRebalanceExtensionInitialized(address(setToken), address(_delegatedManager));
+    }
+
+    /**
+     * @dev ONLY OWNER: Initializes AuctionRebalanceExtension to the DelegatedManager and AuctionRebalanceModuleV1 to the SetToken.
+     *
+     * @param _delegatedManager     Instance of the DelegatedManager to initialize
+     */
+    function initializeModuleAndExtension(IDelegatedManager _delegatedManager) external onlyOwnerAndValidManager(_delegatedManager){
+        ISetToken setToken = _delegatedManager.setToken();
```
Comment (pblivin0x) [inline contracts/global-extensions/GlobalAuctionRebalanceExtension.sol:100]:
```suggestion
        require(_delegatedManager.isPendingExtension(address(this)), "Extension must be pending");

        ISetToken setToken = _delegatedManager.setToken();
```

Comment (snake-poison) [reply contracts/global-extensions/GlobalAuctionRebalanceExtension.sol:105]: implemented in 936ccf8

File: contracts/global-extensions/GlobalAuctionRebalanceExtension.sol (hunk 4 above)
Comment (snake-poison) [reply contracts/global-extensions/GlobalAuctionRebalanceExtension.sol:81]: implemented in 936ccf8

File: contracts/global-extensions/GlobalAuctionRebalanceExtension.sol (hunk 3 above)
Comment (snake-poison) [reply contracts/global-extensions/GlobalAuctionRebalanceExtension.sol:91]: implemented in 936ccf8

File: contracts/global-extensions/GlobalAuctionRebalanceExtension.sol (hunk 2 above)
Comment (snake-poison) [reply contracts/global-extensions/GlobalAuctionRebalanceExtension.sol]: implemented in 936ccf8

File: contracts/global-extensions/GlobalAuctionRebalanceExtension.sol (hunk 1 above)
Comment (snake-poison) [reply contracts/global-extensions/GlobalAuctionRebalanceExtension.sol:59]: implemented in 936ccf8

File: contracts/global-extensions/GlobalAuctionRebalanceExtension.sol [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    /* ============ State Variables ============ */
+    
+    IAuctionRebalanceModuleV1 public immutable auctionModule;  // AuctionRebalanceModuleV1
+
+
+    /* ============ Constructor ============ */
+
+    constructor(IManagerCore _managerCore, IAuctionRebalanceModuleV1 _auctionModule) public BaseGlobalExtension(_managerCore) {
```
Comment (pblivin0x) [inline contracts/global-extensions/GlobalAuctionRebalanceExtension.sol:67]: nit: formatting and natspec to match 
https://github.com/IndexCoop/index-coop-smart-contracts/blob/936ccf83bfd75e7b52b4df231bd8356216d6f6ac/contracts/global-extensions/GlobalWrapExtension.sol#L55-L69

**Review by pblivin0x — APPROVED** [review]
LGTM 💯

**Pull request merged**